### PR TITLE
single-request-response version of spawnAdapter

### PR DESCRIPTION
**Pull request by patriknw**
Alternative impl of https://github.com/akka/akka/pull/24242 using FunctionRef (via spawnAdapter)

Refs #23770

### CODE REVIEW

File: akka-actor-typed/src/main/scala/akka/actor/typed/scaladsl/ActorContext.scala [hunk 1]
```diff
@@ -144,14 +147,43 @@ trait ActorContext[T] { this: akka.actor.typed.javadsl.ActorContext[T] ⇒
...
+   *                      the other actor can send a message back through.
+   * @param responseToOwnProtocol Transforms the response from the `otherActor` into a message this actor understands.
+   *                              Can touch immutable state to provide a context for the interaction, an id for example, but
+   *                              must not touch the `ActorContext` as it will not be executed inside of this actor.
+   *                              Will be fed a `Failure(AskTimeoutException)` if the response timeout hits.
+   *
+   * @tparam Req The request protocol, what the other actor accepts
+   * @tparam Res The response protocol, what the other actor sends back
+   */
+  def ask[Req, Res](
```
Comment (johanandren) [inline akka-actor-typed/src/main/scala/akka/actor/typed/scaladsl/ActorContext.scala]: Document ordering/thread guarantees now that we have them? Also, the warning about not touching the state/ctx isn't true anymore now, right?

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorContextAdapter.scala [hunk 2]
```diff
@@ -60,7 +60,11 @@ import akka.actor.typed.Behavior.UntypedBehavior
   }
   override private[akka] def internalSpawnAdapter[U](f: U ⇒ T, _name: String): ActorRef[U] = {
     val cell = untyped.asInstanceOf[akka.actor.ActorCell]
-    val ref = cell.addFunctionRef((_, msg) ⇒ untyped.self ! f(msg.asInstanceOf[U]), _name)
+    val ref = cell.addFunctionRef((_, msg) ⇒ {
+      val adaptedMsg = f(msg.asInstanceOf[U])
+      // possible to discard by returning null
+      if (adaptedMsg != null) untyped.self ! adaptedMsg
```
Comment (johanandren) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorContextAdapter.scala]: I can't see that null is returned anywhere? Add test case to document how it is used (when you want to ignore the response I guess?)

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/ActorContextImpl.scala [hunk 3]
```diff
@@ -67,5 +74,50 @@ import scala.concurrent.ExecutionContextExecutor
...
+  // Scala impl
+  override def ask[Req, Res](otherActor: ActorRef[Req], createMessage: ActorRef[Res] ⇒ Req)(responseToOwnProtocol: Try[Res] ⇒ T)(implicit timeout: Timeout): Unit = {
+
+    // note that the adapters are running _in_ this actor so there is no concurrency
+
+    var replyTo: ActorRef[Res] = null
+    var reqMsgClassName: String = null
+
+    val timeoutDone: ActorRef[Done] = spawnAdapter { _ ⇒
+      stop(replyTo)
```
Comment (johanandren) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/ActorContextImpl.scala]: Could replyTo still be null here if system is slow and timeout short?

**Review by johanandren — COMMENTED** [review]
Some smaller comments in addition to the question where it actually runs.

**Comment by patriknw** [comment]
I think I got that completely wrong, it's run on the caller of `!`. I have to think some more about this.
Also, since it's actually different ActorRefs for different requests we don't have the ordering guarantee (in Artery).

**Pull request closed without merging**